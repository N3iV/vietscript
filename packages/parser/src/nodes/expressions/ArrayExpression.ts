import { Parser } from "@parser/parser";
import { Literal } from "@parser/nodes/literals/Literal";

export class ArrayExpression {
  type = "ArrayExpression";

  elements: Array<Literal["value"]>;

  constructor(parser: Parser) {
    const elements: Array<Literal["value"]> = [];

    parser.eat("[");

    while (parser.lookahead?.type !== "]") {
      elements.push(new Literal(parser).value);

      if (parser.lookahead?.type !== "]") {
        parser.eat(",");
      }
    }

    parser.eat("]");

    this.elements = elements;
  }
}
