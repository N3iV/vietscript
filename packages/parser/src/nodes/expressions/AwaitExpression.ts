import { Parser } from "@parser/parser";

import { Expression } from "./Expression";

export class AwaitExpression {
  type = "AwaitExpression";

  argument: Expression;

  constructor(parser: Parser) {
    parser.eat("Await");

    this.argument = new Expression(parser);
  }
}
