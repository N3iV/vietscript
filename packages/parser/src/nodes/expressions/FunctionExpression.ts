import { Parser } from "@parser/parser";
import { BlockStatement } from "@parser/nodes/statements/BlockStatement";
import { Identifier } from "@parser/nodes/identifier/Identifier";

export class FunctionExpression {
  type = "FunctionExpression";

  id: null;

  expression: boolean;

  generator: boolean;

  async: boolean;

  params: Array<Identifier>;

  body: BlockStatement;

  constructor(parser: Parser, ignoreKeyword = false, isDefaultAsync = false) {
    let isAsync = isDefaultAsync;
    let isGenerator = false;

    if (parser.lookahead?.type === "Async" && !isDefaultAsync) {
      parser.eat("Async");
      isAsync = true;
    }

    if (!ignoreKeyword) {
      parser.eat("Function");
    }

    if (parser.lookahead?.type === "*") {
      parser.eat("*");
      isGenerator = true;
    }

    parser.eat("(");

    const parameters: Array<Identifier> = [];

    while (parser.lookahead?.type !== ")") {
      if (parameters.length > 0) {
        parser.eat(",");
      }

      parameters.push(new Identifier(parser));
    }

    parser.eat(")");

    const body = new BlockStatement(parser);

    this.id = null;
    this.expression = false;
    this.generator = isGenerator;
    this.async = isAsync;
    this.params = parameters;
    this.body = body;
  }
}
