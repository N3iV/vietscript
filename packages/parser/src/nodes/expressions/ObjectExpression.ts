import { Identifier } from "@parser/nodes/identifier/Identifier";
import { MethodDefinition } from "@parser/nodes/declarations/class/MethodDefinition";
import { PropertyDefinition } from "@parser/nodes/declarations/class/PropertyDefinition";
import { Parser } from "@parser/parser";

export class ObjectExpression {
  type = "ObjectExpression";

  body: Array<MethodDefinition | PropertyDefinition> = [];

  constructor(parser: Parser) {
    parser.eat("{");

    while (parser.lookahead?.type !== "}") {
      const isGetter = parser.lookahead?.type === "Get";

      if (isGetter) {
        parser.eat("Get");
      }
      const isSetter = parser.lookahead?.type === "Set";

      if (isSetter) {
        parser.eat("Set");
      }

      let isComputed = false,
        isAsync = false,
        identifier: Identifier | null = null;

      if (!isGetter && !isSetter) {
        isAsync = parser.lookahead?.type === "Async";

        if (isAsync) {
          parser.eat("Async");
        }

        isComputed = parser.lookahead?.type === "[";

        if (isComputed) {
          parser.eat("[");

          identifier = new Identifier(parser);

          parser.eat("]");
        }
      }

      if (identifier === null) {
        identifier = new Identifier(parser);
      }

      if (parser.lookahead?.type === "(") {
        this.body.push(
          new MethodDefinition(
            parser,
            false,
            isAsync,
            isComputed,
            identifier,
            isGetter ? "get" : isSetter ? "set" : "method",
          ),
        );
      } else {
        this.body.push(new PropertyDefinition(parser, false, isComputed, identifier, ":"));
      }

      if (parser.lookahead?.type !== "}") {
        parser.eat(",");
      }
    }

    parser.eat("}");
  }
}
