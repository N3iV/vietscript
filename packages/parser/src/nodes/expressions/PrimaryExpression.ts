import { Parser } from "@lang/parser";

import { Expression } from "./Expression";

// https://262.ecma-international.org/13.0/#prod-PrimaryExpression
export class PrimaryExpression {
  constructor(parser: Parser) {
    switch (parser.lookahead?.type as string) {
      default: {
        Object.assign(this, new Expression(parser));
      }
    }
  }
}
