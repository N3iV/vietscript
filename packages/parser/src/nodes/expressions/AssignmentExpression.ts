import { Parser } from "@parser/parser";
import { Identifier } from "@parser/nodes/identifier/Identifier";
import { Expression } from "@parser/nodes/expressions/Expression";

export class AssignmentExpression {
  type = "AssignmentExpression";

  left: Identifier | Expression;

  operator: string;

  right: Identifier | Expression;

  constructor(parser: Parser, identifier?: Identifier | Expression) {
    this.left =
      identifier ??
      (parser.lookahead?.type === "Identifier" ? new Identifier(parser) : new Expression(parser));

    this.operator = String(parser.eat("=").value);

    this.right =
      parser.lookahead?.type === "Identifier" ? new Identifier(parser) : new Expression(parser);
  }
}
