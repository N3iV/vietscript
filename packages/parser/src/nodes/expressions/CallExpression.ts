import { Parser } from "@parser/parser";
import { Identifier } from "@parser/nodes/identifier/Identifier";
import { Expression } from "@parser/nodes/expressions/Expression";

export class CallExpression {
  type = "CallExpression";

  callee: Identifier | Expression;

  arguments: Array<Identifier | Expression> = [];

  optional = false;

  constructor(parser: Parser, identifier: Identifier | Expression) {
    this.callee = identifier;

    parser.eat("(");

    while (parser.lookahead?.type !== ")") {
      if (this.arguments.length > 0) {
        parser.eat(",");
      }

      this.arguments.push(new Expression(parser));
    }

    parser.eat(")");
  }
}
