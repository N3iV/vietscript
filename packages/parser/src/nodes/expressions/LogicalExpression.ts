import { Parser } from "@parser/parser";
import { Identifier } from "@parser/nodes/identifier/Identifier";
import { Expression } from "@parser/nodes/expressions/Expression";
import { Keyword } from "@vietscript/shared";

export class LogicalExpression {
  type = "LogicalExpression";

  left: Identifier | Expression;

  operator: string;

  right: Identifier | Expression;

  constructor(parser: Parser, identifier?: Identifier) {
    this.left =
      identifier ?? (parser.lookahead?.type === Keyword.IDENTIFIER ? new Identifier(parser) : new Expression(parser));

    switch (parser.lookahead?.type) {
      case "??":
      case "||":
      case "&&": {
        this.operator = String(parser.eat(parser.lookahead?.type).value);
        break;
      }
    }

    this.right = parser.lookahead?.type === Keyword.IDENTIFIER ? new Identifier(parser) : new Expression(parser);
  }
}
