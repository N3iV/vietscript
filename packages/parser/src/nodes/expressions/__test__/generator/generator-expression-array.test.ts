import generate from "@babel/generator";
import { ArrayExpression } from "@parser/nodes/expressions/ArrayExpression";

import parser from "../../../../setup-test";

describe("generator-expression-array.test", () => {
  it("should parse the syntax normally", () => {
    const code = `[1, "a", đúng, sai, "đúng", "true", NaN]`;

    const ast = parser.parse(code, ArrayExpression);

    const result = generate(ast);

    expect(result.code).toMatchSnapshot();
  });
});
