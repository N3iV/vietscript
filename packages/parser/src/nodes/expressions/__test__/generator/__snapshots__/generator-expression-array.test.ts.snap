// Vitest Snapshot v1

exports[`generator-expression-array.test > should parse the syntax normally 1`] = `"[1, \\"a\\", true, false, \\"đúng\\", \\"true\\", NaN]"`;
